rfcs/widget-restructure-rfc-1.md

```diff
@@ -0,0 +1,554 @@
+}
+
+// vec.rs
+
+fn render(&mut self, mut context: KayakContextRef) {
+	let tree = context.create_tree();
+  
+  for (index, item) in self.data.iter().enumerate() {
+      context.add_widget(item, index, &tree);
+  }
+  
+  context.commit_tree(tree);
+}
+```
+
+##### WidgetTreeBuilder
+
+Alternatively, we could create a `WidgetTreeBuilder` struct that is use specifically for doing everything as above. The benefit of using the builder is that we can move some of the render specific logic to its own separate entity (again separating the concerns). It could be generated from `KayakContextRef::create_tree` and have a similar API:
+
+```rust
+struct WidgetTreeBuilder<'a> {
+  pub tree: WidgetTree,
+  context: &'a mut KayakContext,
+  current_id: Index,
+}
+
+impl<'a> WidgetTreeBuilder<'a> {
+	pub fn add_children(&mut self, children: Children) -> &mut WidgetTreeBuilder {
+    // ...
+  }
+  
+  pub fn add_widget<W: Widget>(&mut self, widget: W, widget_index: usize) -> &mut WidgetTreeBuilder {
+    // ...
+  }
+  
+  pub fn commit(self) {
+    // ...
+  }
+}
+```
+
+> Not super sold on the name of `WidgetTreeBuilder` since it's not really building anything?
+
+<br />
+
+<br />
+
+#### Constructed Widgets
+
+A constructed widget, for the purposes of this RFC, is one that is manually defined like above but uses some traits and derives to allow it to be used as a tagged (`</>`) component. Like the manually defined widgets, it's good for adding customization and extending the widget itself with other functionality.
+
+To integrate with the `rsx` tagging syntax without becoming a fully "functional widget," it must implement `WidgetConstructor` and define its props.
+
+```rust
+trait WidgetConstructor {
+  type Props;
+  fn create_props(base_props: BaseProps) -> Self::Props;
+  fn construct(id: Index, children: Children, props: Self::Props) -> Self;
+}
+
+/// A struct containing the base props most widgets expect.
+/// The widget can choose to use them or not.
+struct BaseProps {
+  styles: Option<Style>,
+  on_event: Option<OnEvent>
+}
+```
+
+> This trait can't be used as a trait object (`dyn WidgetConstructor`), but that's okay. It's only for ensuring the methods exist and are consistent.
+
+By having this bound on the widget struct, the `rsx` macro can then build the widget by doing something like:
+
+```rust
+let mut props = #widget_ident::create_props(base_props);
+#( props.#prop_ident = #prop_value );*
+let widget = #widget_ident::construct(widget_id, children, props);
+```
+
+Generated out might look something like:
+
+```rust
+// <MyWidget foo={Some(1)} bar={"Hello".to_string()} />
+let mut props = MyWidget::create_props(base_props);
+props.foo = Some(1);
+props.bar = "Hello".to_string();
+let widget = MyWidget::construct(widget_id, children, props);
+```
+
+As for the implementation of `Widget:;render`, it will look largely the same as the manual widgets. The main difference is that now props are no longer found directly on the struct.
```

Comment:
```suggestion
As for the implementation of `Widget::render`, it will look largely the same as the manual widgets. The main difference is that now props are no longer found directly on the struct.
```

Reply:
Lol my keyboard is a bit broken. You may see double letters sometimes as well 😅